src/app/CommandHandler.cpp

```diff
@@ -500,15 +552,48 @@ CHIP_ERROR CommandHandler::AddClusterSpecificFailure(const ConcreteCommandPath &
     return AddStatusInternal(aCommandPath, StatusIB(Status::Failure, aClusterStatus));
 }
 
+CHIP_ERROR CommandHandler::PrepareCommand(const ConcreteCommandPath & aRequestCommandPath, const ConcreteCommandPath & aCommandPath,
+                                          bool aStartDataStruct)
+{
+    // Return prematurely in case of requests targeted to a group that should not add command for response purposes.
+    VerifyOrReturnValue(!IsGroupRequest(), CHIP_NO_ERROR);
+    auto * commandRefTableEntry = mCommandRefLookupTable.Find(aRequestCommandPath);
+    return PrepareCommand(commandRefTableEntry, aCommandPath, aStartDataStruct);
+}
+
 CHIP_ERROR CommandHandler::PrepareCommand(const ConcreteCommandPath & aCommandPath, bool aStartDataStruct)
 {
+    // Return prematurely in case of requests targeted to a group that should not add command for response purposes.
+    VerifyOrReturnValue(!IsGroupRequest(), CHIP_NO_ERROR);
+
+    // Legacy code is calling the deprecated version of PrepareCommand. If we are in a case where
+    // there was a single command in the request, we can just assume this response is triggered by
+    // the single command.
+    auto * commandRefTableEntry = mCommandRefLookupTable.GetCommandRefTableEntryIfSingleRequest();
+
+    // At this point application supports Batch Invoke Commands since there does not exist a single command in table,
+    // but application is calling the depricated PrepareCommand. We have no way to determine the associated CommandRef
+    // to put into the InvokeResponse.
+    VerifyOrDieWithMsg(commandRefTableEntry, DataManagement,
+                       "Seemingly device supports batch commands, but is calling the deprecated PrepareCommand API");
+    return PrepareCommand(commandRefTableEntry, aCommandPath, aStartDataStruct);
+}
+
+CHIP_ERROR CommandHandler::PrepareCommand(const CommandRefLookupTable::CommandRefTableEntry * apCommandRefTableEntry,
+                                          const ConcreteCommandPath & aCommandPath, bool aStartDataStruct)
+{
+    VerifyOrReturnError(apCommandRefTableEntry, CHIP_ERROR_INCORRECT_STATE);
     ReturnErrorOnFailure(AllocateBuffer());
 
     mInvokeResponseBuilder.Checkpoint(mBackupWriter);
+    mBackupState = mState;
     //
     // We must not be in the middle of preparing a command, or having prepared or sent one.
     //
-    VerifyOrReturnError(mState == State::Idle, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mState == State::Idle || mState == State::AddedCommand, CHIP_ERROR_INCORRECT_STATE);
```

Comment:
AddedCommand state is only valid if we previously had a command added. Not sure it's 100% needed to check that.

Reply:
This is very much needed. Otherwise we fail to append any more than 1 response. `State::AddedCommand` is where we are if we have successfully added an entry to `InvokeResponses`